### PR TITLE
Re-introduce "CSDS Implementation"

### PR DESCRIPTION
**Pull request by lidizheng**
Related #25038 #25745

The first CSDS attempt #25038 was rolled back because Envoy protos are not considered "core" components, or included in any components. But technically, we use generated upb files which means we depend on Envoy protos transitively anyway.

The most correct solution is adding Envoy protos to "core" component. But there are 2 blockers:

1. Adding Envoy proto packages means adding hundreds of new paths to the "core" component list, which includes Envoy's proto folder, protoc-gen-validate's proto folder, opencensus's proto folder, googleapis's proto folder;
2. Blaze's `compatible_with` tag "non_prod" is not instrumented for Envoy protos, which means we have to update the entire build chain's `compatible_with` tag, including: **envoy**, **protoc-gen-validate**, **opencensus**, **googleapis**, and part of internal **ProtoBuf** ("non_prod" is a superset of "gce"/"mobile", but ironically as a compatible tag it isn't compatible with existing tags) (all above packages are compatible with "gce", so there is no technical reason they won't be compatible with "non_prod").

With the release date set as next Tuesday, I don't think we can clear this two blockers quick enough.

Thanks to @gnossen. I just learnt that our staging test spawns GCE, pulls gRPC from GitHub, and test the OSS version. This means **we can temporarily hide the CSDS feature in google3**, without impacting the ability to test CSDS in production and in staging.

The internal counterpart of this PR is [cl/363946493](http://cl/cl/363946493).
~~[cl/363758897](http://cl/363758897), the example produced import CL looks like [cl/363759740](http://cl/363759740)~~.

### CODE REVIEW

**Comment by markdroth** [comment]
FYI, there's already been some work on moving the xDS protos into core components internally.  See b/169268726 for details.

I'll take a look at the CL internally.

**Comment by lidizheng** [comment]
@markdroth PTALA. I have updated the generic hack with a more specific hack.

File: bazel/grpc_build_system.bzl [hunk 1]
```diff
@@ -128,6 +128,11 @@ def grpc_cc_library(
         linkstatic = linkstatic,
     )
 
+def grpc_cc_library_xds(
```
Comment (markdroth) [inline bazel/grpc_build_system.bzl:131]: Please add a comment here saying that once we can depend on the xDS protos internally, this should go away, and all build rules using it should be changed back to use `grpc_cc_library`.

File: bazel/grpc_build_system.bzl [hunk 2]
```diff
@@ -242,6 +247,9 @@ def grpc_cc_test(name, srcs = [], deps = [], external_deps = [], args = [], data
         **args
     )
 
+def grpc_cc_test_xds(*args, **kwargs):
```
Comment (markdroth) [inline bazel/grpc_build_system.bzl:250]: Same here.

**Review by markdroth — APPROVED** [review]
This looks great!  Please make sure Nico approved before merging.

File: bazel/grpc_build_system.bzl (hunk 1 above)
Comment (lidizheng) [reply bazel/grpc_build_system.bzl:132]: Done.

**Comment by lidizheng** [comment]
@nicolasnoble PTAL to see if the added Bazel rule is okay.

**Comment by lidizheng** [comment]
cl/364351425 is reviewed by Nico. This PR is cherry-picked as cl/364390547. Merging.

**Pull request merged**